pseudocode-defns/pseudocode_defns.tex

```diff
@@ -0,0 +1,525 @@
+        \item \textbf{64-bit machines:} if \texttt{v} is a value of type \texttt{usize}, then $\texttt{v}\in \{0,1,2,\ldots,2^{64} - 1\}$
+    \end{itemize}
+    \inRust{usize}{https://doc.rust-lang.org/std/primitive.usize.html}
+\end{definition}
+
+\begin{definition}[\texttt{Vec(T)}]$\sfi$
+    The Rust type \texttt{Vec(T)} consists of ordered lists of type \texttt{T}. For example, if \texttt{T = bool}, then values of type \texttt{Vec(T)} include $\texttt{[], [0], [1], [0, 0],} \ldots$. \inRust{Vec<T>}{https://doc.rust-lang.org/std/vec/struct.Vec.html}
+\end{definition}
+
+\horizline
+
+\subsection{Notes, todos, questions}
+
+%\silvia{Discuss: bounded vs lower bound and upper bound}
+
+\todonei{Include info on MPFR, and possibly relate it to our existing definitions of floats.}
+
+\todonei{Define plus, minus, etc. below each type on which they operate. For example, the definition for \texttt{u32} should also include a definition of plus on \texttt{u32}, multiplication on \texttt{u32}, etc.}
+
+\questionr{Should we have a general definition for ``floats'' (and ``integers''?), or is it sufficiently understood what a float is in general?}
+
+\section{Domains}
+
+$\sfi$ A \emph{data domain} is a representation of the set of values on which a metric or function can operate. For example, if a function accepts inputs from the domain \texttt{IntervalDomain(1:u32,17:u32)}, this means that the function can take any input value \texttt{v} of type \texttt{u32} such that \texttt{1 <= v} and \texttt{v <= 17}.
+
+\begin{definition}[\texttt{AllDomain}]$\sfi$
+\texttt{AllDomain(T)} is the domain of all values of type $\T$. \domainType{AllDomain[T]}
+\end{definition}
+
+For example, \texttt{AllDomain(u32)} is the domain of all values of type \texttt{u32}.
+
+\begin{definition}[\texttt{IntervalDomain}] \sfi
+\texttt{IntervalDomain(L:T, U:T)} is the domain of all values \texttt{v} of type \texttt{T} such that \texttt{L <= v} and \texttt{v <= U}, for a type \texttt{T} that has a total ordering (\texttt{T} has trait \texttt{TotalOrd}) and for values \texttt{L <= U} of type \texttt{T}. \domainType{IntervalDomain[T]}
+    
+Note that, because both \texttt{L} and \texttt{U} are of type \texttt{T}, there is no need to explicitly pass \texttt{T}; the type \texttt{T} can be inferred. \texttt{IntervalDomain} is defined on any type that implements the trait \texttt{TotalOrd}.\footnote{As of June 28, the OpenDP library requires the weaker condition of partial ordering (implements \texttt{PartialOrd}) instead.}
+\end{definition}
+
+For example, \texttt{IntervalDomain(1:u32, 17:u32)} corresponds to a domain that contains all the \texttt{u32} values \texttt{v} such that \texttt{1 <= v} and \texttt{v <= 17}; it has type \texttt{IntervalDomain[u32]}.
+
+\begin{definition}[\texttt{InherentNullDomain}] 
+    \texttt{InherentNullDomain(inner\_domain:D)} is the domain of all values of data domain \texttt{inner\_domain} and \texttt{null} values. \domainType{InherentNullDomain[D]}

```

Comment:
Do we mean InherentNullDomain(inner_domain) = inner_domain \cup \{null\}?  If so, we should just write this.  (Actually, it's probably more precise to write \llbracket T\rrbracket to denote the set of values represented by type T.  Marco Gaboardi may have an opinion on whether it's important to use the latter notation.)

Also should the type of inner_domain be Domain rather than D?

Reply:
What happens if we write InherentNullDomain(InherentNullDomain(inner_domain))?  Does the result have two different null values or only one?  (i.e. is the union an ordinary union or a disjoint union)

Reply:
We can write it as the union. `InherentNullDomain(InherentNullDomain(inner_domain))` would only have one null value, the outer union being a no-op, meaning we could define a deduplicating transformation. Practically speaking, this domain should be unreachable.

Reply:
I just wrote it as the union for now. Marking resolved. Please let me know if anything else should be done.